File: 0445-add-two-numbers-ii/0445-add-two-numbers-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    // make 2 stack
    let s1 = []
    let s2 = []
    while (l1){
        s1.push(l1.val)
        l1 = l1.next
    }
    while (l2){
        s2.push(l2.val)
        l2 = l2.next
    }
    
	// similiar to 2. Add Two Numbers
    let list = new ListNode(0)
    let sum = 0;
    
    while (s1.length !== 0 || s2.length !== 0 || sum > 0){
        sum = sum + (s1.length === 0? 0: s1.pop())
        sum = sum + (s2.length === 0? 0: s2.pop())
        // start to deal with linked list
		// update current node value
        list.val = sum % 10
		// add new head node with carry, head.val could be 1 or 0
		sum = Math.floor(sum/10);
        let head = new ListNode(sum)
		// connect
        head.next = list
		// update the head
        list = head
    }
    if (list.val === 0){
        return list.next
    } else {
        return list
    }
};

File: 0445-add-two-numbers-ii/0445-add-two-numbers-ii.rs
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        if (l1 == null && l2 == null) return null
        if (l1 == null && l2 != null) return l2
        if (l1 != null && l2 == null) return l1
        var q1: Stack<Int> = Stack()
        var q2: Stack<Int> = Stack()
        var (a, b) = l1 to l2
        var sum: Int = 0
        while (a != null) {
            sum += a.`val`
            q1.push(a.`val`)
            a = a.next
        }
        while (b != null) {
            sum += b.`val`
            q2.push(b.`val`)
            b = b.next
        }
        if (sum == 0) return ListNode(0)
        var resultTail: ListNode? = null
        var resultHead: ListNode? = null
        var carry = 0
        while (q1.isNotEmpty() || q2.isNotEmpty()) {
            sum = carry
            if (q1.isNotEmpty()) sum += q1.pop()
            if (q2.isNotEmpty()) sum += q2.pop()
            resultHead = ListNode(sum % 10).apply { next = resultTail }
            resultTail = resultHead
            carry = sum / 10
        }
        if (carry != 0) resultHead = ListNode(carry).apply { next = resultTail }
        return resultHead
    }
}